src/parsers.ts

```diff
@@ -0,0 +1,153 @@
+    return data
+}
+
+export const parseBuffer = (data: unknown, errMsg: ParseErrorReason): Buffer => {
+    validate(isBuffer(data), errMsg)
+    return data
+}
+
+export const parseBufferOfLength = <L extends number>(data: unknown, length: L, errMsg: ParseErrorReason): FixlenBuffer<L> => {
+    validate(isBufferOfLength(data, length), errMsg)
+    return data
+}
+
+export const parseBufferOfMaxLength = <L extends number>(data: unknown, maxLength: L, errMsg: ParseErrorReason): MaxlenBuffer<L> => {
+    validate(isBufferOfMaxLength(data, maxLength), errMsg)
+    return data
+}
+
+/**
+ * Defines a generic parser that takes only the data, it assumes other
+ * arguments of the parser are already bound.
+ * @see createParser
+ */
+export type Parser<T> = (data: unknown) => T
+
+/**
+ * Takes a parsing function and binds its arguments except for the data
+ * returns a parser function that only requires the data to execute
+ * @see Parser
+ */
+export function createParser<L extends number>(paser: (data: unknown, length: L, errMsg: ParseErrorReason) => FixlenBuffer<L>, length: L, errMsg: ParseErrorReason): Parser<FixlenBuffer<L>>
+export function createParser<L extends number>(paser: (data: unknown, maxLength: L, errMsg: ParseErrorReason) => MaxlenBuffer<L>, maxLength: L, errMsg: ParseErrorReason): Parser<MaxlenBuffer<L>>
+export function createParser<L extends number>(paser: (data: unknown, maxLength: L, errMsg: ParseErrorReason) => MaxlenString<L>, maxLength: L, errMsg: ParseErrorReason): Parser<MaxlenString<L>>
+export function createParser<T, A extends any[]>(parser: (data: unknown, ...args: [...A]) => T, ...args: [...A]): Parser<T>
+export function createParser<T, A extends any[]>(parser: (data: unknown, ...args: [...A]) => T, ...args: [...A]): Parser<T> {
+    return (data: unknown) => parser(data, ...args)
+}
+
+export const parseMap = <K, V>(
+    data: unknown,
+    parseKey: Parser<K>,
+    parseValue: Parser<V>,
+    errMsg: ParseErrorReason,
+): Map<K, V> => {
+    validate(isMap(data), errMsg)
+    return new Map(Array.from(data.entries()).map(([key, value]) => [parseKey(key), parseValue(value)]))
+}
+
+export const parseArray = <T>(
+    data: unknown,
+    parseEntry: Parser<T>,
+    errMsg: ParseErrorReason,
+): Array<T> => {
+    validate(isArray(data), errMsg)
+    return data.map((value) => parseEntry(value))
+}
+
+/**
+ * Defines a tuple of parsers each for the individual type from tuple `T`
+ */
+type Parsers<T extends any[]> = { [K in keyof T]: Parser<T[K]> }
+
+/**
+ * Parses the data as an array of length of the provided parsers
+ *
+ * @example
+ *     // returns [123N, -1N]
+ *     parseTuple([123, -1], InvalidDataReason.INVALID, parseUint64, parseInt64)
+ */
+export const parseTuple = <T extends any[]>(data: unknown, errMsg: ParseErrorReason, ...parsers: [...Parsers<T>]): T => {
+    validate(isArray(data), errMsg)
+    validate(data.length <= parsers.length, errMsg)
```

Comment:
Hm I remember having a reason for it, but now I don't remember what it was. But the tests seem to be working even with:
```ts
validate(data.length === parsers.length, errMsg)
```
So I guess I can change it.

Reply:
It may be useful here (if we decide to go with the option I propose): https://github.com/vacuumlabs/cardano-hw-interop-lib/pull/2#discussion_r748741863

Reply:
@KubqoA you changed this to `===` recently, however, I think `<=` may be better, see my previous comment.